subsys/bluetooth/shell/bt.c

```diff
@@ -938,6 +980,87 @@ static int cmd_scan(const struct shell *sh, size_t argc, char *argv[])
 
 	return 0;
 }
+
+static int cmd_scan_filter_set(const struct shell *sh, size_t argc,
+			       char *argv[])
+{
+	for (size_t argn = 1; argn + 1 < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (strcmp(arg, "name") == 0) {
+			const char *name_arg = argv[++argn];
+
+			shell_print(ctx_shell, "name_arg %s", arg);
+
+			if (strlen(name_arg) >= sizeof(scan_filter.name)) {
+				shell_error(ctx_shell,
+					    "Name is too long (max %zu): %s\n",
+					    sizeof(scan_filter.name), name_arg);
+				return -ENOEXEC;
+			}
+
+			strcpy(scan_filter.name, name_arg);
+			scan_filter.name_set = true;
+		} else if (strcmp(arg, "addr") == 0) {
+			const char *addr_arg = argv[++argn];
+
+			/* Validate length */
+			if (strlen(addr_arg) > sizeof(scan_filter.addr)) {
+				shell_error(ctx_shell,
+					    "Invalid address string: %s\n",
+					    addr_arg);
+				return -ENOEXEC;
+			}
+
+			/* Validate input to check if valid (subset of) BT address */
+			for (size_t i = 0; i < strlen(addr_arg); i++) {
+				const char c = addr_arg[i];
+				uint8_t tmp;
+
+				if (c != ':' && char2hex(c, &tmp) < 0) {
+					shell_error(ctx_shell,
+						    "Invalid address string: %s\n",
+						    addr_arg);
+					return -ENOEXEC;
+				}
+			}
+
+			strcpy(scan_filter.addr, addr_arg);
+			scan_filter.addr_set = true;
+		} else {
+			shell_help(sh);
+			return SHELL_CMD_HELP_PRINTED;
+		}
+	}
+
+	return 0;
+}
+
+static int cmd_scan_filter_clear(const struct shell *sh, size_t argc,
+				 char *argv[])
+{
+	for (size_t argn = 1; argn < argc; argn++) {
+		const char *arg = argv[argn];
+
+		if (strcmp(arg, "all") == 0) {
```

Comment:
Just a proposal. You can replace all the `strcmp` functions with subcommands. In such case handler: `cmd_scan_filter_clear` and `cmd_scan_filter_set` are not needed at all.

Reply:
Not a bad idea. I see how that can eliminate the `strcmp`, but I am not sure how they will eliminate the need of the two `cmd_scan_filter_*` functions. Can you elaborate?

Reply:
Sure.
I was thinking to get rid of these two functions at all. They are not needed. Next, you can implement the following handlers and subcommands:

```
static int cmd_sub_filter_set_name(const struct shell *sh, size_t argc,
				   char *argv[])
{
	const char *name_arg = argv[1];

	shell_print(ctx_shell, "name_arg %s", arg);

	if (strlen(name_arg) >= sizeof(scan_filter.name)) {
		shell_error(ctx_shell,
			    "Name is too long (max %zu): %s\n",
			    sizeof(scan_filter.name), name_arg);
		return -ENOEXEC;
	}

	strcpy(scan_filter.name, name_arg);
	scan_filter.name_set = true;

	return 0;
}

static int cmd_sub_filter_set_addr(const struct shell *sh, size_t argc,
				   char *argv[])
{
	const char *addr_arg = argv[1];

	/* Validate length */
	if (strlen(addr_arg) > sizeof(scan_filter.addr)) {
		shell_error(ctx_shell,
			    "Invalid address string: %s\n",
			    addr_arg);
		return -ENOEXEC;
	}

	/* Validate input to check if valid (subset of) BT address */
	for (size_t i = 0; i < strlen(addr_arg); i++) {
		const char c = addr_arg[i];
		uint8_t tmp;

		if (c != ':' && char2hex(c, &tmp) < 0) {
			shell_error(ctx_shell,
				    "Invalid address string: %s\n",
				    addr_arg);
			return -ENOEXEC;
		}
	}

	strcpy(scan_filter.addr, addr_arg);
	scan_filter.addr_set = true;

	return 0;
}

SHELL_STATIC_SUBCMD_SET_CREATE(sub_scan-filter-set,
	SHELL_CMD_ARG(addr, NULL, "help string addr",
		      cmd_sub_filter_set_addr, 2, 0),
	SHELL_CMD_ARG(name, NULL, "help string addr",
		      cmd_sub_filter_set_name, 2, 0),
	SHELL_SUBCMD_SET_END);

static int cmd_sub_filter_clear_all(const struct shell *sh, size_t argc,
				   char *argv[])
{
	(void)memset(&scan_filter, 0, sizeof(scan_filter));

	return 0;
}

static int cmd_sub_filter_clear_addr(const struct shell *sh, size_t argc,
				   char *argv[])
{
	(void)memset(scan_filter.addr, 0, sizeof(scan_filter.addr));
	scan_filter.addr_set = false;

	return 0;
}

static int cmd_sub_filter_clear_name(const struct shell *sh, size_t argc,
				   char *argv[])
{
	(void)memset(scan_filter.name, 0, sizeof(scan_filter.name));
	scan_filter.name_set = false;

	return 0;
}

SHELL_STATIC_SUBCMD_SET_CREATE(sub_scan-filter-clear,
	SHELL_CMD_ARG(all, NULL, "help string all",
		      cmd_sub_filter_clear_all, 1, 0),
	SHELL_CMD_ARG(addr, NULL, "help string addr",
		      cmd_sub_filter_clear_addr, 1, 0),
	SHELL_CMD_ARG(name, NULL, "help string addr",
		      cmd_sub_filter_clear_name, 1, 0),
	SHELL_SUBCMD_SET_END);
```

These subcommands can be used in your commands like below. I have removed `cmd_scan_filter_set` and `cmd_scan_filter_clear` from below code. These handlers are not needed as long as subcommands have handlers. What is more if one will call this command (despite it has no handler) with wrong arguments shell will pop an error.

```
SHELL_CMD_ARG(scan-filter-set, &sub_scan-filter-set,
		      "[name <name>] [addr <addr>]",
		      NULL, 3, 0),  
	SHELL_CMD_ARG(scan-filter-clear, &sub_scan-filter-clear,
		      "[all] [name <name>] [addr <addr>]",
		      NULL, 2, 0),
```
When implemented this way you can prompt `all`, `addr` and `all` using the tab key.

I do not know if such a solution works for you. I apologize but I did not compile my code. Please treat it more like a "pseudo code".

Reply:
Ah, I understand now. 

Yeah, I did actually have the same thought when I started this PR, but decided to not pursue it. You have convinced me that this is a better approach, and I'll implement it soon :)

Reply:
Refactored to your proposal :)